src/Raven.Server/Indexing/VoronBufferedInput.cs

```diff
@@ -0,0 +1,264 @@
+        buffer = newBuffer;
+        ArrayPool<byte>.Shared.Return(oldBuffer);
+    }
+
+    private void OpenInternal(Transaction transaction)
+    {
+        if (transaction.IsWriteTransaction == false)
+        {
+            if (transaction.LowLevelTransaction.ImmutableExternalState is IndexTransactionCache cache)
+            {
+                if (cache.DirectoriesByName.TryGetValue(_directory.Name, out var files))
+                {
+                    if (files.ChunksByName.TryGetValue(_name, out var details))
+                    {
+                        // we don't dispose here explicitly, the fileName needs to be
+                        // alive as long as the transaction is
+                        Slice.From(transaction.Allocator, _name, out Slice fileName);
+                        _stream = new VoronStream(fileName, details, transaction.LowLevelTransaction);
+                        return;
+                    }
+                }
+            }
+        }
+
+        var fileTree = transaction.ReadTree(_tree);
+        if (fileTree == null)
+            throw new FileNotFoundException($"Could not find '{_tree}' tree for index input", _name);
+
+        using (Slice.From(transaction.Allocator, _name, out Slice fileName))
+        {
+            _stream = fileTree.ReadStream(fileName);
+            if (_stream == null)
+                throw new FileNotFoundException("Could not find index input", _name);
+        }
+    }
+
+    public override object Clone(IState s)
+    {
+        var state = s as VoronState;
+        if (state == null)
+        {
+            ThrowStateNullException();
+            return null;
+        }
+
+        ThrowIfDisposed(state);
+
+        var clone = (VoronBufferedInput)base.Clone(s);
+        GC.SuppressFinalize(clone);
+        clone._isOriginal = false;
+
+        clone.OpenInternal(state.Transaction);
+        clone._stream.Position = _stream.Position;
+
+        return clone;
+    }
+
+    protected override void Dispose(bool disposing)
+    {
+        if (_isDisposed)
+            return;
+
+        GC.SuppressFinalize(this);
+
+        if (buffer != null)
+        {
+            ArrayPool<byte>.Shared.Return(buffer);
+        }
+
+        if (_isOriginal == false)
+            return;
+
+        _cts.Cancel();
+        _cts.Dispose();
+
+
+        _isDisposed = true;
+    }
+
+    public override long Length(IState s)
+    {
+        if (s is VoronState)
+        {
+            return _stream.Length;
+        }
+
+        ThrowStateNullException();
+        return 0;
+    }
+
+
+    public override void ReadInternal(byte[] b, int offset, int length, IState s)
+    {
+        var state = s as VoronState;
+        if (state == null)
+        {
+            ThrowStateNullException();
+            return;
+        }
+
+        ThrowIfDisposed(state);
+
+        _stream.UpdateCurrentTransaction(state.Transaction);
+        _stream.ReadEntireBlock(b, offset, length);
```

Comment:
Shouldn't we also implement the buffering here?

Reply:
We are buffering here, the `b` is a reference to a buffer memory (it is called from Lucene)